File: ci/gc-compat/worker_contract.py
#!/usr/bin/env python3
"""usage: worker_contract.py <corpus-checkout> <camp-binary>

THE §14 WORKER-CONTRACT GATE. Nothing else in this repo runs the REAL
gc-role-worker fragment from the corpus against camp's shims. `e2e_corpus.py`
drives a fake agent that speaks camp's OWN CLI; this gate drives the corpus's
OWN 140-line bash claim protocol — its verbs, its inline `python3` json_pick,
its assignee/route comparisons — against `.camp/bin/gc` and `.camp/bin/bd`, and
proves a gc worker CLOSES a gc bead end to end:

    claim (gc hook --claim --json) -> bd show -> bd update gc.outcome -> bd close
      -> re-hook (drains) -> gc runtime drain-ack -> campd REAPS the worker.

The worker LINGERS after drain-ack (`exec sleep 600`) exactly as a real
`claude -p` does (it does not exit on EOF, P3), so campd's drain-ack ->
KillReleased is the ONLY thing that can reap it inside the deadline. A hang is
the failing signal: if that wiring regresses, the worker sleeps past the
deadline and this gate fails.

It also RE-DERIVES ci/gc-compat/fixtures/gc-role-worker.observed.json from the
live fragment and fails on drift, so moving GCPACKS_REF cannot silently change
the measured contract (§10: we commit only our derived facts, never the
fragment's source).
"""

import json
import os
import re
import shutil
import signal
import subprocess
import sys
import tempfile
import time

DEADLINE = 20  # seconds; < the default release_grace (30s) so the grace backstop
# cannot mask a drain-ack->KillReleased regression.

if len(sys.argv) != 3:
    print(__doc__)
    sys.exit(2)

corpus = os.path.abspath(sys.argv[1])
camp_bin = os.path.abspath(sys.argv[2])
here = os.path.dirname(os.path.abspath(__file__))
repo = os.path.dirname(os.path.dirname(here))  # ci/gc-compat -> ci -> repo root
observed_path = os.path.join(here, "fixtures", "gc-role-worker.observed.json")

FRAGMENT_REL = "gascity/roles/template-fragments/gc-role-worker.template.md"
ROUTE = "gc.run-operator"  # a real gc pack agent; the fragment is its shared body


def die(msg):
    print("WORKER-CONTRACT gate FAIL:", msg)
    sys.exit(1)


# --------------------------------------------------------------------------
# python3 is a HARD gc-worker dependency (§6.1) and must be in the container.
# --------------------------------------------------------------------------
dockerfile = os.path.join(repo, "contrib", "docker", "Dockerfile")
if "python3" not in open(dockerfile).read():
    die(
        "python3 is a hard gc-worker dependency (§6.1: every gc pack agent's "
        f"fragment parses hook --claim --json with an inline python3) and must be "
        f"in the runtime image {dockerfile}"
    )


# --------------------------------------------------------------------------
# Render the REAL fragment into a runnable worker: extract its ```bash claim
# block VERBATIM (the load-bearing contract) and wrap it with the close/drain
# steps the fragment's prose prescribes. This IS the renderer (NB1) — not a
# camp `prime`; camp reads a pack agent's prompt raw.
# --------------------------------------------------------------------------
fragment_path = os.path.join(corpus, FRAGMENT_REL)
if not os.path.isfile(fragment_path):
    die(f"fragment not found at {fragment_path}")
fragment_src = open(fragment_path).read()

# Substitute the Go-template placeholders the block does not need but which
# appear in the file, and drop the define/end wrappers.
rendered = (
    fragment_src.replace("{{ .AgentName }}", ROUTE)
    .replace("{{ .TemplateName }}", ROUTE)
)

# The runnable claim protocol is the first ```bash fenced block (the GC_CLAIM
# heredoc). It claims, verifies assignee/route, and prints CLAIMED_BEAD_ID.
blocks = re.findall(r"```bash\n(.*?)\n```", rendered, re.DOTALL)
if not blocks:
    die("no ```bash block in the fragment — its shape changed; re-measure (Task 1)")
claim_block = blocks[0]
if "gc hook --claim --json" not in claim_block or "GC_CLAIM" not in claim_block:
    die("the first bash block is not the GC_CLAIM claim protocol — re-measure")


def worker_script(mode, claim_path):
    # Run the REAL claim block (captures CLAIMED_BEAD_ID from its stdout), then
    # follow the prose: set gc.outcome, close, re-hook (which drains + drain-
    # acks). The claim block is a `bash <<'GC_CLAIM'` heredoc; running it via
    # `sh <file>` (not inlined in a `$(...)`) keeps the heredoc off the outer
    # shell's parser — the Task-1 harness shape.
    fail = mode == "fail"
    close = (
        "bd update \"$WORK_ID\" --set-metadata 'gc.outcome=fail' "
        "--set-metadata 'gc.failure_class=work_error'\n"
        "bd close \"$WORK_ID\" --reason 'work failed'\n"
        if fail
        else "bd update \"$WORK_ID\" --set-metadata 'gc.outcome=pass'\n"
        "bd close \"$WORK_ID\"\n"
    )
    return (
        "#!/bin/sh\nset +e\n"
        f'OUT="$(sh {claim_path})"\n'
        "printf '%s\\n' \"$OUT\"\n"
        "WORK_ID=\"$(printf '%s\\n' \"$OUT\" | sed -n 's/^CLAIMED_BEAD_ID=//p')\"\n"
        # No work claimed (first hook drained) — the block already drain-acked.
        'if [ -z "$WORK_ID" ]; then exit 0; fi\n'
        f"{close}"
        # Continuation: re-run the REAL claim block; the bead is closed now, so
        # it drains and runs `gc runtime drain-ack`, then exits 0.
        f"sh {claim_path}\n"
    )


# --------------------------------------------------------------------------
# Drift guard: re-derive the observed facts from the LIVE fragment and compare
# to the committed fixture (Task 1). A moved GCPACKS_REF that changes the
# contract must update the fixture, not slip through.
# --------------------------------------------------------------------------
observed = json.load(open(observed_path))
static_pairs = sorted(set(re.findall(r"\b(?:gc|bd) [a-z][a-z-]*", fragment_src)))
if static_pairs != sorted(observed["verbs_static"]["pairs"]):
    die(
        "the fragment's static verb set DRIFTED from the committed measurement.\n"
        f"  live:      {static_pairs}\n"
        f"  committed: {sorted(observed['verbs_static']['pairs'])}\n"
        "Re-run Task 1's measurement and update "
        "ci/gc-compat/fixtures/gc-role-worker.observed.json."
    )
for field in observed["hook_json_fields"]["top_level"]:
    if f"json_pick {field}" not in fragment_src and f'json_pick("{field}")' not in fragment_src:
        # the fragment reads these via `| json_pick <field>`
        if re.search(rf"json_pick\s+{re.escape(field)}\b", fragment_src) is None:
            die(f"the fragment no longer reads hook field {field!r} — re-measure")


def run_mode(mode):
    """Drive real campd + a lingering fake-claude running the REAL fragment in
    `mode`; assert closed + drain_acked + reaped within DEADLINE."""
    # A SHORT tempdir: the campd unix socket path must fit SUN_LEN (~104).
    work = tempfile.mkdtemp(prefix=f"wc-{mode}-")
    campd = None
    try:
        subprocess.run(
            [camp_bin, "init", "--no-service", "--no-import"],
            cwd=work, capture_output=True, check=True,
        )
        root = os.path.join(work, ".camp")
        rig = os.path.join(work, "repo")
        os.makedirs(rig, exist_ok=True)
        git_env = {
            **os.environ,
            "GIT_AUTHOR_NAME": "wc", "GIT_AUTHOR_EMAIL": "wc@example.com",
            "GIT_COMMITTER_NAME": "wc", "GIT_COMMITTER_EMAIL": "wc@example.com",
        }
        for argv in (
            ["git", "init", "-q"],
            ["git", "-c", "commit.gpgsign=false", "commit", "-q", "--allow-empty", "-m", "base"],
        ):
            r = subprocess.run(argv, cwd=rig, capture_output=True, text=True, env=git_env)
            if r.returncode != 0:
                die(f"{' '.join(argv)}: {r.stderr.strip()}")

        # The fake claude: run the rendered REAL fragment, then LINGER. The
        # REAL claim block lives in its own file (run via `sh`), verbatim.
        claim_sh = os.path.join(work, "claim_block.sh")
        open(claim_sh, "w").write(claim_block + "\n")
        worker_sh = os.path.join(work, "worker.sh")
        open(worker_sh, "w").write(worker_script(mode, claim_sh))
        fake = os.path.join(work, "fake-claude.sh")
        open(fake, "w").write(f"#!/bin/sh\nsh {worker_sh} 1>&2\nexec sleep 600\n")
        os.chmod(fake, 0o755)

        with open(os.path.join(root, "camp.toml"), "a") as fh:
            fh.write(
                f'\n[[rigs]]\nname = "gc"\npath = "{rig}"\nprefix = "gc"\n\n'
                f'[agent_defaults]\ntools = ["Read", "Bash"]\n\n'
                f'[dispatch]\nmax_workers = 4\ncommand = "{fake}"\n'
            )

        def camp(*argv, check=True):
            r = subprocess.run([camp_bin, "--camp", root, *argv], capture_output=True, text=True)
            if check and r.returncode != 0:
                die(f"camp {' '.join(argv)} exited {r.returncode}: {r.stderr.strip()}")
            return r

        # Import the gc role pack (the deployment recipe, §3/§7.3) so
        # `gc.run-operator` resolves and the fragment's route check passes.
        camp("import", "add", os.path.join(corpus, "gascity", "roles"), "--name", "gc")

        # start_new_session: campd leads its own process group, so cleanup can
        # killpg the WHOLE tree — campd AND the lingering `sleep 600` worker it
        # spawned (which inherits the group). NB3: no orphaned 600s sleep.
        campd = subprocess.Popen(
            [camp_bin, "--camp", root, "daemon"],
            stdout=subprocess.PIPE, stderr=subprocess.DEVNULL,
            env={**os.environ, "CAMP_BIN": camp_bin},
            start_new_session=True,
        )
        campd.stdout.readline()  # the socket line: campd is listening

        def events(kind=None):
            out = camp("events", "--json").stdout
            evs = [json.loads(l) for l in out.splitlines() if l.strip()]
            return [e for e in evs if kind is None or e["type"] == kind]

        r = camp("sling", "a real gc bead", "--agent", ROUTE)
        bead = r.stdout.split()[0] if r.stdout.split() else "gc-1"

        start = time.time()
        while True:
            evs = events()
            closed = any(e["type"] == "bead.closed" and e.get("bead") == bead for e in evs)
            drain_acked = any(e["type"] == "worker.drain_acked" for e in evs)
            reaped = any(e["type"] == "session.stopped" for e in evs)
            failed = [e for e in evs if e["type"] == "dispatch.failed"]
            crashed = [e for e in evs if e["type"] == "session.crashed"]
            if failed:
                die(f"[{mode}] campd failed a dispatch: "
                    f"{[e.get('data', {}).get('reason') for e in failed]}")
            if crashed:
                die(f"[{mode}] a worker crashed: {[e['data'] for e in crashed]}")
            if closed and drain_acked and reaped:
                return
            if time.time() - start > DEADLINE:
                die(
                    f"[{mode}] campd did not reap the drained gc worker within "
                    f"{DEADLINE}s: the drain-ack->KillReleased wiring regressed, or the "
                    f"real fragment hung against the shims. closed={closed} "
                    f"drain_acked={drain_acked} reaped={reaped}"
                )
            time.sleep(0.1)
    finally:
        if campd is not None:
            subprocess.run([camp_bin, "--camp", os.path.join(work, ".camp"), "stop"],
                           capture_output=True)
            try:
                campd.wait(timeout=10)
            except subprocess.TimeoutExpired:
                campd.kill()
            # Reap the whole group — the lingering `sleep 600` worker included.
            try:
                os.killpg(campd.pid, signal.SIGKILL)
            except ProcessLookupError:
                pass
        shutil.rmtree(work, ignore_errors=True)


run_mode("happy")
run_mode("fail")
print(
    "WORKER-CONTRACT gate OK: the REAL gc-role-worker fragment claimed, closed, "
    "and drain-acked a gc bead against camp's shims on BOTH the happy and "
    "fail-close branches; campd reaped the lingering worker via drain-ack in each "
    "(a hang would have failed). Static verb set matches the committed measurement."
)

File: crates/camp/tests/worker_contract.rs
//! THE §14 unskippable gate (hermetic Rust half).
//!
//! A fixture camp (real ledger, real shims installed by `launch`, a fake claude
//! in place of the model) drives the worker fragment claim → close → drain-ack
//! → REAP under a wall-clock deadline. A hang is the failing signal.
//!
//! Two properties:
//!  1. The byte-projection (§6.1): the hook JSON, `bd show --json`, and the
//!     worker env all project the ONE bead row. The genuinely independent leg
//!     is env `GC_AGENT` vs the cooked route — pinned here by driving the REAL
//!     shim binaries with `GC_AGENT` set to a DIFFERENT value than the bead's
//!     route and proving the shims still read the route from the BEAD (B1/NB1).
//!  2. The lifecycle (§6.2): the worker LINGERS after drain-ack (a real
//!     `claude -p` does not exit on EOF, P3), so campd's drain-ack →
//!     KillReleased is what reaps it. If that wiring regressed, the worker
//!     sleeps past the deadline and the watchdog fails RED (R2-B2).

use std::io::{BufRead, BufReader};
use std::os::unix::process::CommandExt as _;
use std::path::{Path, PathBuf};
use std::process::{Child, Command, Stdio};
use std::time::{Duration, Instant};

const BIN: &str = env!("CARGO_BIN_EXE_camp");
const READY_PREFIX: &str = "campd listening on ";
/// Deadline < the DEFAULT release_grace (30s) ON PURPOSE (fold-in NB2): the
/// grace backstop must NOT be able to mask a drain-ack→KillReleased regression.
/// A green reap inside 20s can therefore only be the drain-ack prompt kill.
const DEADLINE: Duration = Duration::from_secs(20);

fn camp(root: &Path, args: &[&str]) -> std::process::Output {
    Command::new(BIN)
        .env_remove("CAMP_DIR")
        .arg("--camp")
        .arg(root)
        .args(args)
        .output()
        .unwrap()
}

fn camp_ok(root: &Path, args: &[&str]) -> String {
    let out = camp(root, args);
    assert!(
        out.status.success(),
        "camp {args:?}: {}",
        String::from_utf8_lossy(&out.stderr)
    );
    String::from_utf8(out.stdout).unwrap()
}

fn events_json(root: &Path) -> Vec<serde_json::Value> {
    camp_ok(root, &["events", "--json"])
        .lines()
        .map(|l| serde_json::from_str(l).unwrap())
        .collect()
}

/// A camp whose `[dispatch].command` is `fake_claude`, with a `dev` agent and a
/// `gc` rig. Uses the DEFAULT release_grace (30s) — deliberately > DEADLINE.
fn scaffold(dir: &Path, fake_claude: &Path) -> PathBuf {
    let root = dir.join(".camp");
    std::fs::create_dir_all(&root).unwrap();
    let rig = dir.join("repo");
    std::fs::create_dir_all(&rig).unwrap();
    std::fs::write(
        root.join("camp.toml"),
        format!(
            "[camp]\nname = \"t\"\n\n[[rigs]]\nname = \"gc\"\npath = \"{}\"\nprefix = \"gc\"\n\n\
             [agent_defaults]\ntools = [\"Read\", \"Bash\"]\n\n\
             [dispatch]\nmax_workers = 4\ncommand = \"{}\"\ndefault_agent = \"dev\"\n",
            rig.display(),
            fake_claude.display(),
        ),
    )
    .unwrap();
    let dev = root.join("agents/dev");
    std::fs::create_dir_all(&dev).unwrap();
    std::fs::write(dev.join("agent.toml"), "isolation = \"none\"\n").unwrap();
    std::fs::write(dev.join("prompt.md"), "Do the work.\n").unwrap();
    camp_ok(&root, &["events", "--json"]); // create the ledger
    root
}

/// Write the fake claude: run the fragment against the real shims, then LINGER
/// (`exec sleep 600`) — the wrapper stays alive exactly as `claude -p` does, so
/// campd's KillReleased is the only thing that can reap it inside the deadline.
fn write_fake_claude(dir: &Path, fragment: &Path, mode: &str) -> PathBuf {
    let path = dir.join("fake-claude.sh");
    std::fs::write(
        &path,
        format!(
            "#!/bin/sh\n# claude-style argv is ignored; the contract arrives in CAMP_* env.\n\
             GC_FRAGMENT_MODE={mode} sh {} 1>&2\nexec sleep 600\n",
            fragment.display()
        ),
    )
    .unwrap();
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt as _;
        std::fs::set_permissions(&path, std::fs::Permissions::from_mode(0o755)).unwrap();
    }
    path
}

fn fragment_fixture() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR")).join("tests/fixtures/gc-fragment.sh")
}

/// campd in its OWN process group, so cleanup can reap the whole tree — campd
/// AND the lingering `sleep 600` worker (which inherits the group). Fold-in
/// NB3: a failing test must not orphan a 600s sleep.
struct Daemon {
    child: Child,
    pgid: i32,
}

impl Daemon {
    fn spawn(root: &Path, envs: &[(&str, &str)]) -> Daemon {
        let mut cmd = Command::new(BIN);
        cmd.env_remove("CAMP_DIR")
            .env("CAMP_BIN", BIN)
            .args(["daemon", "--camp"])
            .arg(root)
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::inherit())
            .process_group(0); // new group; pgid == campd pid
        for (k, v) in envs {
            cmd.env(k, v);
        }
        let mut child = cmd.spawn().unwrap();
        let pgid = child.id() as i32;
        let stdout = child.stdout.take().unwrap();
        let mut line = String::new();
        BufReader::new(stdout).read_line(&mut line).unwrap();
        assert!(
            line.starts_with(READY_PREFIX),
            "unexpected first line from campd: {line:?}"
        );
        Daemon { child, pgid }
    }
}

impl Drop for Daemon {
    fn drop(&mut self) {
        // Kill the WHOLE group: campd and any worker it spawned (the lingering
        // sleep 600). `kill -KILL -<pgid>` needs no libc/nix dependency.
        let _ = Command::new("kill")
            .arg("-KILL")
            .arg(format!("-{}", self.pgid))
            .status();
        let _ = self.child.kill();
        let _ = self.child.wait();
    }
}

// ---------------------------------------------------------------------------
// Step 1 — the byte projection, with a real env≠bead mismatch (B1/NB1).
// ---------------------------------------------------------------------------

#[test]
fn hook_bd_show_and_env_project_the_same_bead_row() {
    let dir = tempfile::tempdir().unwrap();
    let root = scaffold(dir.path(), Path::new("/bin/true"));
    // A routed bead: `sling --agent dev` stamps beads.assignee = "dev" (the
    // route → gc.routed_to). No campd runs here, so sling exits nonzero ("not
    // dispatched") AFTER durably creating the bead — read its id from the row.
    let _ = camp(&root, &["sling", "do a thing", "--agent", "dev"]);
    let bead = events_json(&root)
        .into_iter()
        .find(|e| e["type"] == "bead.created")
        .expect("sling creates the bead durably even without campd")["bead"]
        .as_str()
        .unwrap()
        .to_owned();
    assert_eq!(bead, "gc-1");

    // Drive the REAL hook shim with GC_AGENT set to a WRONG route on purpose.
    let hook = Command::new(BIN)
        .arg("--camp")
        .arg(&root)
        .args(["gc-shim", "hook", "--claim", "--json"])
        .env("CAMP_BEAD", "gc-1")
        .env("CAMP_SESSION", "t/dev/1")
        .env("GC_AGENT", "gc.WRONG") // <-- ≠ the cooked route "dev"
        .env("GC_TEMPLATE", "gc.WRONG")
        .output()
        .unwrap();
    assert!(hook.status.success(), "hook: {}", String::from_utf8_lossy(&hook.stderr));
    let hook_json: serde_json::Value =
        serde_json::from_slice(&hook.stdout).expect("hook prints one JSON object");
    assert_eq!(hook_json["action"], "work");
    assert_eq!(hook_json["assignee"], "t/dev/1"); // the session
    assert_eq!(hook_json["route"], "dev"); // the BEAD's route — NOT gc.WRONG

    // And bd show projects the SAME row (top-level assignee = the session;
    // metadata.gc.routed_to = the route), both from readiness::bead_metadata.
    let show = Command::new(BIN)
        .arg("--camp")
        .arg(&root)
        .args(["bd-shim", "show", "gc-1", "--json"])
        .env("GC_AGENT", "gc.WRONG")
        .output()
        .unwrap();
    assert!(show.status.success(), "bd show: {}", String::from_utf8_lossy(&show.stderr));
    let bd_json: serde_json::Value = serde_json::from_slice(&show.stdout).unwrap();
    assert_eq!(bd_json["assignee"], hook_json["assignee"]); // same session
    assert_eq!(bd_json["metadata"]["gc.routed_to"], hook_json["route"]); // same route
    assert_eq!(bd_json["metadata"]["gc.routed_to"], "dev"); // NOT env gc.WRONG
}

// ---------------------------------------------------------------------------
// Step 2 — the lingering-worker loop with a real watchdog.
// ---------------------------------------------------------------------------

/// Drive real campd + a fake claude running the fragment in `mode`; assert the
/// bead closes, the worker drain-acks, AND campd reaps the LINGERING worker —
/// all inside DEADLINE. Because the worker `exec sleep 600`s, a reap inside 20s
/// can only be campd's KillReleased; a drain-ack→KillReleased regression sleeps
/// past the deadline → the watchdog panics RED.
fn run_fragment_and_expect_reap(mode: &str) {
    let dir = tempfile::tempdir().unwrap();
    let fake = write_fake_claude(dir.path(), &fragment_fixture(), mode);
    let root = scaffold(dir.path(), &fake);
    let _campd = Daemon::spawn(&root, &[]);

    let bead = camp_ok(&root, &["sling", "do a thing", "--agent", "dev"])
        .trim()
        .to_owned();

    let start = Instant::now();
    loop {
        let events = events_json(&root);
        let closed = events
            .iter()
            .any(|e| e["type"] == "bead.closed" && e["bead"] == bead.as_str());
        let drain_acked = events.iter().any(|e| e["type"] == "worker.drain_acked");
        let reaped = events.iter().any(|e| e["type"] == "session.stopped");
        if closed && drain_acked && reaped {
            // Exactly the §6.2 chain: the bead closed, the worker acked, and
            // campd's KillReleased reaped the still-sleeping worker in time.
            return;
        }
        if start.elapsed() > DEADLINE {
            panic!(
                "campd did not reap the drained worker within {DEADLINE:?} \
                 (mode={mode}): the drain-ack→KillReleased wiring regressed, or the \
                 fragment hung. closed={closed} drain_acked={drain_acked} reaped={reaped}\n\
                 events: {events:#?}"
            );
        }
        std::thread::sleep(Duration::from_millis(50));
    }
}

#[test]
fn campd_reaps_a_lingering_gc_worker_via_drain_ack_after_it_closes_the_bead() {
    run_fragment_and_expect_reap("happy");
}

#[test]
fn the_fail_close_branch_also_closes_and_is_reaped_without_a_hang() {
    run_fragment_and_expect_reap("fail");
}
